cpp/src/strings/regex/regcomp.cpp

```diff
@@ -570,182 +570,180 @@ class regex_parser {
           break;
-        case QUEST:
-          op2                                    = popand('?');
-          id_inst1                               = m_prog.add_inst(OR);
-          id_inst2                               = m_prog.add_inst(NOP);
-          m_prog.inst_at(id_inst1).u2.left_id    = id_inst2;
-          m_prog.inst_at(id_inst1).u1.right_id   = op2.id_first;
-          m_prog.inst_at(op2.id_last).u2.next_id = id_inst2;
-          pushand(id_inst1, id_inst2);
+        }
+        case QUEST: {
+          auto const op                        = pop_and();
+          auto const id_inst1                  = _prog.add_inst(OR);
+          auto const id_inst2                  = _prog.add_inst(NOP);
+          _prog.inst_at(id_inst1).u2.left_id   = id_inst2;
+          _prog.inst_at(id_inst1).u1.right_id  = op.id_first;
+          _prog.inst_at(op.id_last).u2.next_id = id_inst2;
+          push_and(id_inst1, id_inst2);
           break;
-        case QUEST_LAZY:
-          op2                                    = popand('?');
-          id_inst1                               = m_prog.add_inst(OR);
-          id_inst2                               = m_prog.add_inst(NOP);
-          m_prog.inst_at(id_inst1).u2.left_id    = op2.id_first;
-          m_prog.inst_at(id_inst1).u1.right_id   = id_inst2;
-          m_prog.inst_at(op2.id_last).u2.next_id = id_inst2;
-          pushand(id_inst1, id_inst2);
+        }
+        case QUEST_LAZY: {
+          auto const op                        = pop_and();
+          auto const id_inst1                  = _prog.add_inst(OR);
+          auto const id_inst2                  = _prog.add_inst(NOP);
+          _prog.inst_at(id_inst1).u2.left_id   = op.id_first;
+          _prog.inst_at(id_inst1).u1.right_id  = id_inst2;
+          _prog.inst_at(op.id_last).u2.next_id = id_inst2;
+          push_and(id_inst1, id_inst2);
           break;
+        }
       }
     }
   }
 
-  void Operator(int t)
+  void handle_operator(int token, int subid = 0)
   {
-    if (t == RBRA && --nbra < 0)
+    if (token == RBRA && --_bracket_count < 0) {
       // unmatched right paren
       return;
-    if (t == LBRA) {
-      nbra++;
-      if (lastwasand) Operator(CAT);
-    } else
-      evaluntil(t);
-    if (t != RBRA) pushator(t);
-    lastwasand = (t == STAR || t == QUEST || t == PLUS || t == STAR_LAZY || t == QUEST_LAZY ||
-                  t == PLUS_LAZY || t == RBRA);
+    }
+    if (token == LBRA) {
+      _bracket_count++;
+      if (_last_was_and) { handle_operator(CAT, subid); }
+    } else {
+      eval_until(token);
+    }
+    if (token != RBRA) { push_operator(token, subid); }
+
+    static std::vector<int> tokens{STAR, STAR_LAZY, QUEST, QUEST_LAZY, PLUS, PLUS_LAZY, RBRA};
+    _last_was_and =
+      std::any_of(tokens.cbegin(), tokens.cend(), [token](auto t) { return t == token; });
   }
 
-  void Operand(int t)
+  void handle_operand(int token, int subid = 0, char32_t yy = 0, int class_id = 0)
```

Comment:
Is `yy` a reserved word in regex? Could we use a more descriptive name here? The same for https://github.com/rapidsai/cudf/blob/6a64ce13538fd041c3735ad99609fd9bcaf41b1a/cpp/src/strings/regex/regcomp.cpp#L158-L161

Reply:
This PR is mainly for the `regex_compiler`. I'm cleaning up the `regex_parser` in this PR #10975

Reply:
Got it.